crates/sui-graphql-rpc/src/types/epoch.rs

```diff
@@ -210,35 +221,44 @@ impl Epoch {
 
     /// Look up an `Epoch` in the database, optionally filtered by its Epoch ID. If no ID is
     /// supplied, defaults to fetching the latest epoch.
-    pub(crate) async fn query(db: &Db, filter: Option<u64>) -> Result<Option<Self>, Error> {
+    pub(crate) async fn query(
+        db: &Db,
+        filter: Option<u64>,
+        checkpoint_viewed_at: Option<u64>,
+    ) -> Result<Option<Self>, Error> {
         use epochs::dsl;
 
         let id = filter.map(|id| id as i64);
-        let stored: Option<QueryableEpochInfo> = db
-            .execute(move |conn| {
-                conn.first(move || {
-                    let mut query = dsl::epochs
-                        .select(QueryableEpochInfo::as_select())
-                        .order_by(dsl::epoch.desc())
-                        .into_boxed();
-
-                    if let Some(id) = id {
-                        query = query.filter(dsl::epoch.eq(id));
-                    }
-
-                    query
-                })
-                .optional()
+        let (stored, checkpoint_viewed_at): (Option<QueryableEpochInfo>, u64) = db
+            .execute_repeatable(move |conn| {
+                let checkpoint_viewed_at = match checkpoint_viewed_at {
+                    Some(value) => Ok(value),
+                    None => Checkpoint::available_range(conn).map(|(_, rhs)| rhs),
+                }?;
+
+                let stored = conn
+                    .first(move || {
+                        let mut query = dsl::epochs
+                            .select(QueryableEpochInfo::as_select())
+                            .order_by(dsl::epoch.desc())
+                            .into_boxed();
+
+                        if let Some(id) = id {
```

Comment:
Should we be applying a filter based on `checkpoint_viewed_at` here? In particular, it would seem like the interpretation of `filter = None` ("latest epoch") should be altered by that.

Reply:
Thanks for catching. Added filtering logic as follows:
                        // Bound the query on `checkpoint_viewed_at` by filtering for the epoch
                        // whose `first_checkpoint_id <= checkpoint_viewed_at`, selecting the epoch
                        // with the largest `first_checkpoint_id` among the filtered set.